crates/turbopack-core/src/source_map/mod.rs

```diff
@@ -320,6 +478,63 @@ impl SectionedSourceMap {
     pub fn new(sections: Vec<SourceMapSection>) -> Self {
         Self { sections }
     }
+
+    pub async fn source_contents(&self) -> Result<Vec<(String, Option<String>)>> {
+        let mut sources = IndexMap::new();
+
+        for section in &self.sections {
+            let map = section.map.await?;
+            for (source, contents) in map.source_contents().await? {
+                match sources.entry(source) {
+                    indexmap::map::Entry::Occupied(_) => {}
+                    indexmap::map::Entry::Vacant(e) => {
+                        e.insert(contents);
+                    }
+                }
+            }
+        }
+
+        Ok(sources
+            .iter()
+            .map(|(k, v)| (k.to_string(), v.clone()))
+            .collect())
+    }
+
+    #[async_recursion::async_recursion]
+    pub async fn tokens(&self) -> Result<Vec<Token>> {
+        let mut tokens = vec![];
+        for section in &self.sections {
+            let section_tokens = &*section.map.await?.tokens().await?;
+
+            for token in section_tokens {
+                // Derived from https://github.com/getsentry/rust-sourcemap/blob/f1b758a251a5edddc0767f58f8391912c062c889/src/types.rs#L946
+                let generated_line = token.generated_line() + section.offset.line;
+                let generated_column = token.generated_column()
+                    + if token.generated_line() == 0 {
+                        section.offset.column
+                    } else {
+                        0
+                    };
+
+                tokens.push(match token {
```

Comment:
We need to know if this token is out-of-range (it crosses into the next section), and we can skip processing the rest of the tokens in this section if so. The correct implementation is https://github.com/jridgewell/trace-mapping/blob/1abee9ca32e18f075da802a015019118fcfe9484/src/any-map.ts#L48-L155, or an optimal implementation at https://github.com/jridgewell/trace-mapping/blob/8af9820e129906ca7ef29e58a7fe196e1d671c6d/src/any-map.ts#L49-L158